File: protocol-autoresearch/experiments/tier2_006_sdk_designer_proof_envelope_codec/sdk/src/envelope.ts
/**
 * BolyraEnvelope — self-describing proof envelope with named public signals.
 *
 * Replaces bare { proof, publicSignals: bigint[] } with a versioned,
 * typed envelope that maps signal names to values.
 */

import {
  type CircuitName,
  type ProvingSystem,
  SIGNAL_MAPS,
  VALID_PROVING_SYSTEMS,
} from './circuits/signal-maps.js';

export { type CircuitName, type ProvingSystem } from './circuits/signal-maps.js';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

export const ENVELOPE_VERSION = '1.0.0';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface SnarkProof {
  readonly pi_a: string[];
  readonly pi_b: string[][];
  readonly pi_c: string[];
  readonly protocol: string;
  readonly curve: string;
}

export interface BolyraEnvelope {
  readonly version: string;
  readonly circuit: CircuitName;
  readonly provingSystem: ProvingSystem;
  readonly signals: Record<string, string>;
  readonly proof: SnarkProof;
}

export interface DecodedEnvelope {
  readonly version: string;
  readonly circuit: CircuitName;
  readonly provingSystem: ProvingSystem;
  readonly signals: Record<string, string>;
  readonly proof: SnarkProof;
  /** Ordered bigint array matching the circuit's positional signal layout. */
  readonly publicSignals: bigint[];
}

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

export class EnvelopeVersionError extends Error {
  constructor(version: string) {
    super(`Unsupported envelope version: ${version}`);
    this.name = 'EnvelopeVersionError';
  }
}

export class UnknownCircuitError extends Error {
  constructor(circuit: string) {
    super(`Unknown circuit: ${circuit}`);
    this.name = 'UnknownCircuitError';
  }
}

export class SignalCountMismatch extends Error {
  constructor(circuit: string, expected: number, got: number) {
    super(`Signal count mismatch for ${circuit}: expected ${expected}, got ${got}`);
    this.name = 'SignalCountMismatch';
  }
}

export class InvalidProvingSystemError extends Error {
  constructor(circuit: string, provingSystem: string) {
    super(`Invalid proving system '${provingSystem}' for circuit '${circuit}'`);
    this.name = 'InvalidProvingSystemError';
  }
}

// ---------------------------------------------------------------------------
// encode()
// ---------------------------------------------------------------------------

/**
 * Encode a raw proof and positional public signals into a BolyraEnvelope.
 *
 * @param circuit - Circuit that produced the proof
 * @param provingSystem - Proving system used (groth16 or plonk)
 * @param rawProof - Opaque proof object from snarkjs
 * @param rawSignals - Positional public signals array (string-encoded bigints)
 * @returns Versioned BolyraEnvelope with named signals
 */
export function encode(
  circuit: CircuitName,
  provingSystem: ProvingSystem,
  rawProof: SnarkProof,
  rawSignals: string[],
): BolyraEnvelope {
  const signalNames = SIGNAL_MAPS[circuit];
  if (!signalNames) {
    throw new UnknownCircuitError(circuit);
  }

  const validSystems = VALID_PROVING_SYSTEMS[circuit];
  if (!validSystems.includes(provingSystem)) {
    throw new InvalidProvingSystemError(circuit, provingSystem);
  }

  if (rawSignals.length !== signalNames.length) {
    throw new SignalCountMismatch(circuit, signalNames.length, rawSignals.length);
  }

  const signals: Record<string, string> = {};
  for (let i = 0; i < signalNames.length; i++) {
    signals[signalNames[i]] = rawSignals[i];
  }

  return {
    version: ENVELOPE_VERSION,
    circuit,
    provingSystem,
    signals,
    proof: rawProof,
  };
}

// ---------------------------------------------------------------------------
// decode()
// ---------------------------------------------------------------------------

/**
 * Decode and validate a BolyraEnvelope, producing ordered publicSignals[].
 *
 * @param envelope - Raw envelope object (e.g. parsed from JSON)
 * @returns DecodedEnvelope with both named signals and positional bigint array
 */
export function decode(
  envelope: Record<string, unknown>,
): DecodedEnvelope {
  const version = envelope.version as string;
  if (!version || typeof version !== 'string') {
    throw new EnvelopeVersionError(String(version));
  }

  const majorVersion = parseInt(version.split('.')[0], 10);
  if (isNaN(majorVersion) || majorVersion > 1) {
    throw new EnvelopeVersionError(version);
  }

  const circuit = envelope.circuit as string;
  if (!(circuit in SIGNAL_MAPS)) {
    throw new UnknownCircuitError(circuit);
  }
  const circuitName = circuit as CircuitName;

  const provingSystem = envelope.provingSystem as string;
  const validSystems = VALID_PROVING_SYSTEMS[circuitName];
  if (!validSystems.includes(provingSystem as ProvingSystem)) {
    throw new InvalidProvingSystemError(circuitName, provingSystem);
  }

  const signalNames = SIGNAL_MAPS[circuitName];
  const signals = envelope.signals as Record<string, string>;
  if (!signals || typeof signals !== 'object') {
    throw new SignalCountMismatch(circuitName, signalNames.length, 0);
  }

  const signalKeys = Object.keys(signals);
  if (signalKeys.length !== signalNames.length) {
    throw new SignalCountMismatch(circuitName, signalNames.length, signalKeys.length);
  }

  const publicSignals: bigint[] = [];
  for (const name of signalNames) {
    if (!(name in signals)) {
      throw new SignalCountMismatch(circuitName, signalNames.length, signalKeys.length);
    }
    publicSignals.push(BigInt(signals[name]));
  }

  return {
    version,
    circuit: circuitName,
    provingSystem: provingSystem as ProvingSystem,
    signals,
    proof: envelope.proof as SnarkProof,
    publicSignals,
  };
}

// ---------------------------------------------------------------------------
// fromRaw() — migration helper
// ---------------------------------------------------------------------------

/**
 * Migration helper: wrap a raw snarkjs proof output into a BolyraEnvelope.
 *
 * This is a thin alias for encode() that serves as the upgrade path for
 * integrators currently using positional publicSignals[] arrays.
 *
 * @example
 * ```ts
 * // Before (positional):
 * const { proof, publicSignals } = await snarkjs.groth16.fullProve(...);
 * const root = publicSignals[0]; // fragile positional indexing
 *
 * // After (named):
 * const envelope = fromRaw('HumanUniqueness', 'groth16', proof, publicSignals);
 * const root = envelope.signals.humanMerkleRoot; // self-documenting
 * ```
 */
export function fromRaw(
  circuit: CircuitName,
  provingSystem: ProvingSystem,
  rawProof: SnarkProof,
  rawSignals: string[],
): BolyraEnvelope {
  return encode(circuit, provingSystem, rawProof, rawSignals);
}

File: protocol-autoresearch/experiments/tier2_006_sdk_designer_proof_envelope_codec/sdk-python/bolyra/envelope.py
"""BolyraEnvelope — self-describing proof envelope with named public signals.

Python mirror of the TypeScript envelope codec. Produces and consumes
the same JSON wire format so envelopes are cross-SDK compatible.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

# ---------------------------------------------------------------------------
# Signal maps (must match sdk/src/circuits/signal-maps.ts)
# ---------------------------------------------------------------------------

HUMAN_UNIQUENESS_SIGNALS = (
    "humanMerkleRoot",
    "nullifierHash",
    "nonceBinding",
)

AGENT_POLICY_SIGNALS = (
    "credentialCommitment",
    "permissionsBitmask",
    "scopeCommitment",
    "expiryTimestamp",
)

DELEGATION_SIGNALS = (
    "delegatorCredCommitment",
    "delegateeCredCommitment",
    "narrowedPermissionsBitmask",
    "delegationNullifier",
)

SIGNAL_MAPS: dict[str, tuple[str, ...]] = {
    "HumanUniqueness": HUMAN_UNIQUENESS_SIGNALS,
    "AgentPolicy": AGENT_POLICY_SIGNALS,
    "Delegation": DELEGATION_SIGNALS,
}

VALID_PROVING_SYSTEMS: dict[str, tuple[str, ...]] = {
    "HumanUniqueness": ("groth16",),
    "AgentPolicy": ("groth16", "plonk"),
    "Delegation": ("groth16", "plonk"),
}

ENVELOPE_VERSION = "1.0.0"

# ---------------------------------------------------------------------------
# Errors
# ---------------------------------------------------------------------------


class EnvelopeVersionError(ValueError):
    """Raised when the envelope version is unsupported."""


class UnknownCircuitError(ValueError):
    """Raised when the circuit name is not recognised."""


class SignalCountMismatch(ValueError):
    """Raised when signal count does not match the circuit's expected count."""


class InvalidProvingSystemError(ValueError):
    """Raised when the proving system is not valid for the circuit."""


# ---------------------------------------------------------------------------
# Dataclass
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class BolyraEnvelope:
    """Self-describing proof envelope with named public signals."""

    version: str
    circuit: str
    proving_system: str
    signals: dict[str, str]
    proof: dict[str, Any]

    def to_dict(self) -> dict[str, Any]:
        """Serialize to a JSON-compatible dict matching the TS wire format."""
        return {
            "version": self.version,
            "circuit": self.circuit,
            "provingSystem": self.proving_system,
            "signals": dict(self.signals),
            "proof": self.proof,
        }

    def to_public_signals(self) -> list[int]:
        """Reconstruct the ordered positional public signals as ints."""
        signal_names = SIGNAL_MAPS.get(self.circuit)
        if signal_names is None:
            raise UnknownCircuitError(f"Unknown circuit: {self.circuit}")
        return [int(self.signals[name]) for name in signal_names]


# ---------------------------------------------------------------------------
# encode()
# ---------------------------------------------------------------------------


def encode(
    circuit: str,
    proving_system: str,
    raw_proof: dict[str, Any],
    raw_signals: list[str],
) -> BolyraEnvelope:
    """Encode raw proof and positional signals into a BolyraEnvelope."""
    signal_names = SIGNAL_MAPS.get(circuit)
    if signal_names is None:
        raise UnknownCircuitError(f"Unknown circuit: {circuit}")

    valid_systems = VALID_PROVING_SYSTEMS[circuit]
    if proving_system not in valid_systems:
        raise InvalidProvingSystemError(
            f"Invalid proving system '{proving_system}' for circuit '{circuit}'"
        )

    if len(raw_signals) != len(signal_names):
        raise SignalCountMismatch(
            f"Signal count mismatch for {circuit}: "
            f"expected {len(signal_names)}, got {len(raw_signals)}"
        )

    signals = {name: raw_signals[i] for i, name in enumerate(signal_names)}

    return BolyraEnvelope(
        version=ENVELOPE_VERSION,
        circuit=circuit,
        proving_system=proving_system,
        signals=signals,
        proof=raw_proof,
    )


# ---------------------------------------------------------------------------
# decode()
# ---------------------------------------------------------------------------


def decode(data: dict[str, Any]) -> BolyraEnvelope:
    """Decode and validate a BolyraEnvelope from a JSON-compatible dict."""
    version = data.get("version")
    if not isinstance(version, str):
        raise EnvelopeVersionError(f"Unsupported envelope version: {version}")

    parts = version.split(".")
    try:
        major = int(parts[0])
    except (ValueError, IndexError):
        raise EnvelopeVersionError(f"Unsupported envelope version: {version}")

    if major > 1:
        raise EnvelopeVersionError(f"Unsupported envelope version: {version}")

    circuit = data.get("circuit", "")
    if circuit not in SIGNAL_MAPS:
        raise UnknownCircuitError(f"Unknown circuit: {circuit}")

    proving_system = data.get("provingSystem", "")
    valid_systems = VALID_PROVING_SYSTEMS[circuit]
    if proving_system not in valid_systems:
        raise InvalidProvingSystemError(
            f"Invalid proving system '{proving_system}' for circuit '{circuit}'"
        )

    signal_names = SIGNAL_MAPS[circuit]
    signals = data.get("signals", {})
    if not isinstance(signals, dict):
        raise SignalCountMismatch(
            f"Signal count mismatch for {circuit}: "
            f"expected {len(signal_names)}, got 0"
        )

    if len(signals) != len(signal_names):
        raise SignalCountMismatch(
            f"Signal count mismatch for {circuit}: "
            f"expected {len(signal_names)}, got {len(signals)}"
        )

    for name in signal_names:
        if name not in signals:
            raise SignalCountMismatch(
                f"Missing signal '{name}' for circuit {circuit}"
            )

    return BolyraEnvelope(
        version=version,
        circuit=circuit,
        proving_system=proving_system,
        signals=signals,
        proof=data.get("proof", {}),
    )


# ---------------------------------------------------------------------------
# fromRaw() — migration helper
# ---------------------------------------------------------------------------


def from_raw(
    circuit: str,
    proving_system: str,
    raw_proof: dict[str, Any],
    raw_signals: list[str],
) -> BolyraEnvelope:
    """Migration helper: alias for encode().

    Provides a clear upgrade path for integrators using positional arrays.
    """
    return encode(circuit, proving_system, raw_proof, raw_signals)
